{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 471,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Assignment 3 main program \n",
    "#The program does not work...\n",
    "\n",
    "require 'net/http'\n",
    "require 'bio'\n",
    "\n",
    "def fetch(uri_str)  # fetch that does some error-handling\n",
    "  address = URI(uri_str)  # create URI\n",
    "  response = Net::HTTP.get_response(address)  # this is used to call the URI adress\n",
    "  case response   \n",
    "    when Net::HTTPSuccess then  # when response is of type Net::HTTPSuccess\n",
    "      return response  # return that response object\n",
    "    else\n",
    "      raise Exception, \"Something went wrong... the call to #{uri_str} failed; type #{response.class}\"\n",
    "      response = False\n",
    "      return response  # now we are returning False\n",
    "  end \n",
    "end\n",
    "\n",
    "#Using the gene ids, we get the entries\n",
    "  \n",
    "  #Step 1. Array of gene ids\n",
    "genes = File.open(\"ArabidopsisSubNetwork_GeneList.txt\",\"r\")\n",
    "geneids = genes.read.split() #remove newlines -> we get an array\n",
    "geneids = geneids.join\",\" #turns array into a string, each id is separated by a comma\n",
    "genes.close\n",
    "#puts geneids\n",
    "  \n",
    "   #Step 2. Get the entries\n",
    "unless File.file?(\"Genes.embl\") #if this file already exists, use it\n",
    "  ensembl = File.open(\"Genes.embl\", \"w\") #otherwise we create it\n",
    "  uri = \"https://www.ebi.ac.uk/Tools/dbfetch/dbfetch?db=ensemblgenomesgene&format=embl&id=\" + geneids\n",
    "  res = fetch(uri)\n",
    "  data = res.body\n",
    "  ensembl.write(data) #we add the entries of our genes to the embl file\n",
    "  ensembl.close\n",
    "end\n",
    "\n",
    "#Preparing files\n",
    "gff = File.open(\"feats.gff\", \"w\")\n",
    "nox = File.open(\"noexons.txt\", \"w\")  \n",
    "  \n",
    "#Where are the exons? \n",
    "entries = Bio::FlatFile.auto('Genes.embl') #reading the ensembl file: Bioruby autodetects the file format. \n",
    "entries.each do |entry|\n",
    "  unless entry.features[1].nil?\n",
    "    geneid = entry.features[1].assoc[\"gene\"]\n",
    "  end\n",
    "  pos=[] # array to store exon positions\n",
    "  entry.features do |feature|\n",
    "    next unless feature.feature == \"exon\"\n",
    "    pos = pos.push(feature.position) # store all exon positions in the array\n",
    "  end\n",
    "  \n",
    "  seq = entry.naseq #We get the nucleotide sequence of the entry\n",
    "  #name = entry.entry_id\n",
    "  #fasta = seq.to_fasta(name) #method to get fasta format\n",
    "  bioseq = Bio::Sequence::NA.new(seq) #this is a sequence object\n",
    "  pos.each do |exon|\n",
    "    if exon[0].to_i > 0 #the exon is in forward strand. Format: start_pos..end_pos\n",
    "      start,ends = exon.split\"..\" #we get the start and end position\n",
    "      #puts start, ends\n",
    "      d_exon = bioseq.subseq[start.to_i..ends.to_i] #now we have the sequence of the exon\n",
    "=begin\n",
    "we are looking for exons that contain the CTTCTT sequence and we want to have the coordinates-> regular expression time!\n",
    "I used this method to find a match:\n",
    "      search = Bio::Sequence::NA.new(\"CTTCTT\")\n",
    "      re = Regexp.new(search.to_re)\n",
    "      match = d_exon.match(re)\n",
    "BUT I did not know how to get the coordinates of both forward and reverse strands, so I used a method provided by my classmates:\n",
    "=end\n",
    "  forward = Array.new\n",
    "  reverse = Array.new\n",
    "      match = d_exon.enum_for(:scan, /cttctt/).map{Regexp.last_match.begin(0)} \n",
    "      unless match.nil? ##it keeps storing empty positions into the array, I don't know why! :(\n",
    "        match.each do |x|\n",
    "          a = x.to_i + start.to_i \n",
    "          b = a + 5\n",
    "          forward = forward.push([a,b])\n",
    "        end\n",
    "        \n",
    "      match = d_exon.enum_for(:scan, /aagaag/).map{Regexp.last_match.begin(0)} \n",
    "      unless match.nil? \n",
    "        match.each do |x|\n",
    "          a = x.to_i + start.to_i \n",
    "          b = a + 5\n",
    "          reverse = reverse.push([a,b])\n",
    "        end\n",
    "      end\n",
    "      end\n",
    "\n",
    "=begin\n",
    "this code did not work\n",
    "\n",
    "    elsif exon[0].to_s == 'c' #the exon is in the reverse strand. Format: complement(XXXXX:start_pos..end_pos)\n",
    "      complem, pos = exon.split\":\"\n",
    "      pos = pos.delete!')' #Program stops here. NoMethodError: undefined method `delete!' for nil:NilClass\n",
    "      start, ends = pos.split\"..\" #we get the start and end position of the exon\n",
    "      s = start.to_i \n",
    "      e = ends.to_i\n",
    "      d_exon = bioseq.subseq[e-s..bioseq.length.to_i-e-s] #now we have the sequence of the exon\n",
    "      match = d_exon.enum_for(:scan, /aagaag/).map{Regexp.last_match.begin(0)} #since this is the reverse strand, we are looking for the complement of CTTCTT\n",
    "      match.each do |x|\n",
    "        a = x.to_i + s\n",
    "        b = a + 5\n",
    "        reverse = reverse.push(a,b)\n",
    "      end\n",
    "    end\n",
    "=end\n",
    "      #Let's try to get a gff file at least\n",
    "      forward.uniq.each do |pos|\n",
    "        gff.puts\"#{geneid}\\tEMBL\\tCTTCTT\\t#{pos[0]}\\t#{pos[1]}\\t.\\t+\\t\" \n",
    "        f1 = Bio::Feature.new('CTTCTT', \"#{pos[0]}..#{pos[1]}\")\n",
    "        f1.append(Bio::Feature::Qualifier.new('strand', '+'))\n",
    "      end\n",
    "      reverse.uniq.each do |pos|\n",
    "        gff.puts\"#{geneid}\\tEMBL\\tCTTCTT\\t#{pos[0]}\\t#{pos[1]}\\t.\\t-\\t\"\n",
    "        f1 = Bio::Feature.new('CTTCTT', \"#{pos[0]}..#{pos[1]}\")\n",
    "        f1.append(Bio::Feature::Qualifier.new('strand', '-'))\n",
    "      end\n",
    "    end\n",
    "  end\n",
    "end\n",
    "gff.close\n",
    "\n",
    "gff = File.open(\"feats.gff\", \"r\")\n",
    "have_exons = Array.new\n",
    "gff.each do |line|\n",
    "  line = line.split\"\\t\" #line to array\n",
    "  line.each do |text|\n",
    "    text = text.chomp \n",
    "  end\n",
    "  have_exons << line[0] #array of gene ids that have CTTCTT sequence in their exons\n",
    "end\n",
    "\n",
    "gfile = File.open(\"ArabidopsisSubNetwork_GeneList.txt\",\"r\")\n",
    "genenames = gfile.read.split() #remove newlines -> we get an array of all the gene ids\n",
    "genenames = genenames.sort.map(&:upcase) #returns an uppercased array\n",
    "have_exons = have_exons.uniq.sort.map(&:upcase) #same, so we can substract them\n",
    "\n",
    "no_exons = genenames - have_exons\n",
    "no_exons.each do |nexon|\n",
    "  nox.puts nexon\n",
    "end\n",
    "gfile.close\n",
    "nox.close\n",
    "\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 2.3.1",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "2.3.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
